### ISSUE TITLE
PlayGamesSdk.initialize(<application instance>) should be called on Android on application start

### ISSUE THREAD

**Issue by andynewman10** [Abedalkareem/games_services#127]
From what I understand, since `games_services` is using the V2 Play Games SDK, it must call `PlayGamesSdk.initialize` on application start, as specified here:

https://developers.google.com/games/services/android/signin?hl=en#initialize_the_sdk_2

This should be done before calling any other function.

**Comment by Abedalkareem** [comment]
Hi @andynewman10 
Agree I think we should do that, maybe I'll added in the sign in method.

**Comment by theLee3** [comment]
According to the [method summary](https://developers.google.com/android/reference/com/google/android/gms/games/PlayGamesSdk):

> It is only necessary to call this method if you disable the `PlayGamesInitProvider` provider declared by the Play Games SDK. Under normal circumstances `PlayGamesInitProvider` will automatically initialize the Play Games SDK at application startup.

**Comment by andynewman10** [comment]
I can't find any documentation specifying how to disable the PlayGamesInitProvider.

The only thing I could find is this blog entry from August 2022:

https://novalis.org/blog/2022-08-03-dev-diary--surfwords.html

Specifically, this paragraph:

_When I [remove the provider](https://novalis.org/blog/play-games-hack.patch) (using horrible Gradle hackery)..._

**Comment by theLee3** [comment]
@andynewman10 I appreciate the research you're putting in here.

> To be consistent between Android and iOS shouldn't games_services consider disabling PlayGamesInitProvider?

This is where cross-platform development can get tricky. There are 2 approaches to cross-platform development in this respect:
  1. Complete consistency across all devices (usually to maintain branding).
  2. Adhering to the expected look & behavior of each platform so that it feels native.

In this case, as a plugin calling native features on each OS, I believe it is best to adhere to the developer guidelines for each platform to provide the expected native experience on each user's device. It is less about experiencing the same behavior on all devices and more about the user experiencing the same Play Games or Game Center behavior in your game that they experience on most or all games on their device. Given this, the current guidelines state specifically that the user should be automatically signed in on app launch. If they don't want this behavior, they are to adjust the sign-in settings on there device in settings or the Google Play Games app.

**Comment by andynewman10** [comment]
@theLee3 your points are completely valid, I agree.

> @andynewman10 I appreciate the research you're putting in here.
> 
> > To be consistent between Android and iOS shouldn't games_services consider disabling PlayGamesInitProvider?
> 
> This is where cross-platform development can get tricky. There are 2 approaches to cross-platform development in this respect:
> 
> 1. Complete consistency across all devices (usually to maintain branding).
> 2. Adhering to the expected look & behavior of each platform so that it feels native.
> 
> In this case, as a plugin calling native features on each OS, I believe it is best to adhere to the developer guidelines for each platform to provide the expected native experience on each user's device. It is less about experiencing the same behavior on all devices and more about the user experiencing the same Play Games or Game Center behavior in your game that they experience on most or all games on their device.

@theLee3 your points are completely valid, I agree.

> Given this, the current Google Play Games guidelines state specifically that the user should be automatically signed in on app launch. If they don't want this behavior, they are to adjust the sign-in settings on there device in settings or the Google Play Games app.

The only trouble, here, is that Google actually _withdrew_ the possibility to opt out of automatic sign-in, both in the device settings and in the Play Games App. If we consider this and the fact that the `PlayGamesInitProvider` documentation is missing... I find it a bit "weird". Weird to "very weird".

Another thing to consider is that a user may want to choose to sign-in with an alternate method, eg. email/password, Firebase or something else. In this case, signing-in with Play Games feels odd - the user may indeed think this is a bug.

**Comment by andynewman10** [comment]
> It's actually still available, though in making Play Games much more "user friendly" (and "dev friendly"), they seemed to have made it more difficult to find. However, once you know where it is, the granular controls are quite nice.
> 
> In the Play Games App you must select the _Profile_ tab, then select _Settings_ from the overflow menu. Under _Your data_, select _Change account for games_. Here you can set your default account for all games, as well as individually for each game. You are able to select any Google account on the device, _Ask each time_, or _Sign out_.
> 
> The same options can be found in Settings starting with Android 14 by going to _Settings_ > _Google_ > _Settings for Google apps_ > _Play Games_ then selecting _Sign in account_.

How can I thank you enough for pointing this out? Many thanks for your insight. I tried, and I can indeed sign out from the Play Games App.

> 
> > Another thing to consider is that a user may want to choose to sign-in with an alternate method, eg. email/password, Firebase or something else. In this case, signing-in with Play Games feels odd - the user may indeed think this is a bug.
> 
> I believe that your desired behavior should be implemented with the [google_sign_in](https://pub.dev/packages/google_sign_in) package. There is a `SignInOption.games` configuration that would allow the user to sign in with their Play Games account. This option can then be provided alongside other sign-in methods. The games_services plugin is meant to be used exclusively with Play Games and Game Center to provide access to features such as achievements, leaderboards, and game saves. Given that you're willing to allow sign-in via other means, I would assume these are not used or are to be handled separately (i.e. via Firestore).

Thanks, I will look into the `SignInOption.games` option of [google_sign_in](https://pub.dev/packages/google_sign_in).

> If you wish to implement Play Games achievements, etc. if signed in via Play Games, then these two approaches must co-exist in your app. This is not uncommon in mobile games with in-game account sign-in/creation and Play Games sign-in occurring simultaneously.

That's exactly what I'm doing. I want users to be able to sign-in with Play Games, or with an email/password, for instance. If signing-in with Play Games, I do want to be able to publish achievements.

The problem I have and that made me post the current issue is the following scenario:

- a user chooses to sign in Play Games in my game
- play a bit
- then sign-out from the Play Games-bound account in my game "logically" (meaning the sign-out takes place logically for the game only, since there is no `signOut` method in the Play Games SDK)
- The user then closes the application. So far so good.
- he/she now launches the application again, the Play Games popup will show up, and that feels weird: the user may want to forget about Play Games altogether this time, and use email/password credentials. In this situation I can either leave the Play Games popup or, as you pointed out, not use games_services and use google_sign_in instead. But I'll have to forget leaderboards/achievements etc.

> Also of note, Apple's guidelines also state that automatic login should be attempted at app launch, and that if unsuccessful, no button should be provided to login via Game Center. They continue to make the developer responsible for attempting the auto-login, as did Google before Play Games Services v2. Based on your use case, you may wish to use _Sign in with Apple_ via the [sign_in_with_apple](https://pub.dev/packages/sign_in_with_apple) package rather than Game Center.

Thanks again, but this one is for signing in with Apple ID, not Game Center, isn't it? If that's the case, that's not suitable for me. Play Games / Game Center remove friction because they hide email addresses away (whilst providing a unique user id).

**Comment by theLee3** [comment]
So your issue is absolutely due to Google changing the way Play Games auth works with v2. You could continue to use the last version of games_services before we implemented v2, but this will likely seize to work at some point due to Google phasing out v1. Maybe way in the future though. Of note, when using v1, games_services also used google_sign_in under the hood with the SignInOptions.games config as it was actually necessary to request the Google Drive scope for saving games to the cloud.

As you stated, the google_sign_in approach would work best for your sign-in flow, but you would have to handle achievements & leaderboards via API calls. Feasible, but extra work.

You are correct that Apple accounts and Game Center accounts are treated as separate accounts despite one sitting behind the other. So this is not ideal for you. I am curious how you handle your desire for sign out capabilities in app for Game Center though, as this is not possible either. Like with Play Games v2, this is only available via the OS.

**Comment by theLee3** [comment]
However, you do make a fair point. Perhaps the plugin should go out of its way to let the developer ignore the guidelines if possible. We could potentially provide an optional parameter to not initialize on launch. Allowing you to initialize manually later.

Like you though, I had trouble upon first effort learning about `PlayGamesInitProvider`. Even Google's doc [links](https://developers.google.com/android/reference/com/google/android/gms/games/provider/PlayGamesInitProvider) are broken.

**Comment by andynewman10** [comment]
As you noted, the Play Games documentation suggests it is possible to disable PlayGamesInitProvider. So I consider disabling it possible, and official.

This can be done in the following way, we need to add the following block of code in the `android` section of the app's build.gradle file:

```
android.applicationVariants.all { variant -> 
        variant.outputs.each { output ->
            def processManifest = output.getProcessManifestProvider().get()
            processManifest.doLast { task ->
                def outputDir = task.getMultiApkManifestOutputDirectory()
                File outputDirectory
                if (outputDir instanceof File) {
                    outputDirectory = outputDir
                } else {
                    outputDirectory = outputDir.get().asFile
                }
                File manifestOutFile = file("$outputDirectory/AndroidManifest.xml")

                if (manifestOutFile.exists() && manifestOutFile.canRead() && manifestOutFile.canWrite()) {
                    def xmlParser = new XmlParser(false, false)
                    def xmlRoot = xmlParser.parse(manifestOutFile)
                    
                    def providerNode = xmlRoot.'application'.'*'.find { node ->
                        node.name() == 'provider' && node['@android:name'] == 'com.google.android.gms.games.provider.PlayGamesInitProvider'
                    }
                    if (providerNode != null) {
                        println("PlayGamesInitProvider node detected and removed from manifest")
                        providerNode.replaceNode {}
                    }
                    else {
                        println("PlayGamesInitProvider node not detected in manifest")
                    }

                    def nodePrinter = new XmlNodePrinter(new PrintWriter(new FileWriter(manifestOutFile)))
                    nodePrinter.preserveWhitespace = true
                    nodePrinter.print(xmlRoot)
                }
            }
        }
    }
```
I'm using Gradle 7.x.

I modified `games_services` the following way:

```
// Auth.kt

  private var initialized: Boolean = false

  fun isSignedIn(activity: Activity?, result: MethodChannel.Result) {
    if (initialized == false) {
      result.success(false)
      return
    }
   ...
 }

  fun signIn(activity: Activity?, result: MethodChannel.Result) {
    if (initialized == false) {
      Log.i("PlayService", "Initializing Play Games")
      PlayGamesSdk.initialize(activity!!) // internally, initialize() will call getApplicationContext then cast to Application
      initialized = true
    }
    ...
   }
```

Note this is using google-play-services v2 version 19.x (which is the latest version ; I believe games_services uses the previous version, which, by the way, has a reported security vulnerability).

I tested and this works:

- when you launch the application, the init provider is not started
- when you call signIn(), no popup is displayed. This is not ideal, unfortunately, but this still is OK for me.
- if you minimze the application, then bring it back to the foreground, the popup will be displayed (only once). That's a pity, because at this point the popup would best be never displayed at all since it was not displayed when signIn() had first been called.

Tested on Android 7.1.1 using an internal test version of the application published on the store (running a debug sometimes yields slightly different results wrt. the experience with the Play Games popup).

**Comment by andynewman10** [comment]
> Also of note, Apple's guidelines also state that automatic login should be attempted at app launch, and that if unsuccessful, no button should be provided to login via Game Center. They continue to make the developer responsible for attempting the auto-login, as did Google before Play Games Services v2. Based on your use case, you may wish to use _Sign in with Apple_ via the [sign_in_with_apple](https://pub.dev/packages/sign_in_with_apple) package rather than Game Center.

When you call GamesServices.signIn() on iOS and Game Center has not been enabled, if somehow the user rejects the sign-in in the Auth screen that appears (similar to the OAuth screen on Android), Apple says that you should not attempt to try to sign-in again: basically, Apple says the user should not be bothered a second time. And actually - and that's the funny part - if you **do** attempt to call signIn() again, the program will be frozen (stuck in `await`ing forever a response from signIn()). The only way to unfreeze the app in this case is to minimize it and bring it back to the foreground. Maybe I'll file a separate issue for that, but only if this makes sens wrt. the Apple guidelines (so far this second call seem to violate the guidelines).

**Comment by andynewman10** [comment]
> I believe that your desired behavior should be implemented with the [google_sign_in](https://pub.dev/packages/google_sign_in) package. There is a `SignInOption.games` configuration that would allow the user to sign in with their Play Games account. This option can then be provided alongside other sign-in methods. The games_services plugin is meant to be used exclusively with Play Games and Game Center to provide access to features such as achievements, leaderboards, and game saves.

@theLee3 the consent screen for Play Games (`games_services`) and `google_sign_in` are not the same. Although the consent screen typically appears only once when you use your smartphone, the difference is noticeable and is a red light for me.

Can you tell me if this matches your experience? For potential readers of this issue, this might give additional insight about the pros and cons of both methods, which is interesting I think.

My apology, these captures are in French but I believe they give sufficient information and can mostly be understood.

The game name here is displayed as "project-XXX" (second capture), as I have not sent the project for validation yet.

The first capture below is for Play Games. As a reminder for users stumbling upon this thread, this consent screen is presented to the user if, from a game, a Play Games sign-in occurs and you have, eg., never configured/launched Play Games before (as I was saying, obviously it usually happens only once):

![play_games_consent_screen](https://github.com/Abedalkareem/games_services/assets/113537099/a259ba18-c5fd-41df-8bc8-c94edc736860)

The second screen below is the one proposed by `google_sign_in` when `GoogleSignIn.games().signIn()` is called:

![google_sign_in_games_consent_screen](https://github.com/Abedalkareem/games_services/assets/113537099/407d2d2a-b49f-4d80-90fb-73d11ef85b2c)

This is significantly different. The second capture suggests the Google account is "somewhat" exposed, whereas with Play Games, it mostly feels hidden away, invisible, inaccessible (which is good).

`google_sign_in` seems to really be about signing-in to a Google Account, **asking for the specific scope "Play Games"**, whereas `games_services` (Play Games v2 - let's forget about the one) seems to be a dedicated option, with a dedicated consent screen that is not provided google_sign_in.

Two last questions I have for you: could you please tell me if

1) games_services v1 (not the newer one using the Play Games v2 API) triggers a Play Games popup sometimes?
2) google_sign_in's `GoogleSignIn.games().signIn()` causes the appearance of the Play Games popup?

I presume that's not the case for 2), but not sure about 1)?

**Comment by andynewman10** [comment]
I filed an issue:

Play Games popup counter-intuitive behavior when playgamesinitprovider is disabled
https://issuetracker.google.com/issues/313677100

**Comment by andynewman10** [comment]
PlayGamesInitProvider documentation missing issue:
https://issuetracker.google.com/issues/313687373

**Comment by theLee3** [comment]
I also submitted the missing doc issue via documentation feedback.

**Comment by theLee3** [comment]
Silly me should have pointed this out already, but even when using GoogleSignIn, it has to be used alongside Play Games Services. Therefore, even if you use GoogleSignIn, which yes, will always look different, you have to pass the credential on to the PGS SDK to activate Play Games features. Therefore, you still wouldn't get around integrating PGS, which as of v2 will automatically login without disabling the great mystery that is `PlayGamesInitProvider`.

**Comment by andynewman10** [comment]
Thanks again for your insight, I think we're very soon going to reach the bottom of this.

(by Play Games _popup_, I really meant the bubble-style window that appears at the very top of the screen ; the consent screen is the consent given by the user through various half-screen (Play Games v1 consent) or full-screen (Play Games v2 consent) overlay dialogs - just to clarify but I think you got it right!)

Regarding the mysterious `PlayGamesInitProvider`, I looked up the decompiled source code and saw that this provider does its work through an `ActivityLifecycleCallback`: when the application is launched, I suppose the provider starts very slightly before the main activity, and because of that, observes that activity's lifecycle events such as `onResume` when it happens. From there, it displays the Play Games popup. `PlayGamesSdk.initialize` also does exactly this, unfortunately, meaning `onResume` only happens in this case when you minimize the app then bring it back to the foreground. Not sure, but that's my feeling that's what happens.

I still want to do more tests. Right now, I'm struggling to find a web page, bound to a Google account, where I could reset all consents, I mean dismiss them all, to start over my tests from scratch.

**Comment by andynewman10** [comment]
@theLee3 I think I know what is the best approach to adopt.

As you said, it should be up to the end developer to determine if it is yes or no acceptable for a game to be signed-in all the time or not. I thought this out for a long time and it currently seems the best strategy. Play Games v2 does have the PlayGamesInitProvider "issue" with the popup suddenly appearing, and having an always signed-in mode is odd.

Currently, `games_services` supports both "modes".

1) new style, sign-in only mode (with no sign-out option): that's `games_services: 4.x` using Play Games v2 APIs.
2) old-style, sign-in/sign-out mode: that's `games_services: 3.0.3` using Play Games v1 APIs (I also think this API will stick around for a long time)

Regarding 2), there is one thing I am still not clear about. I hope I am not abusing your time.

- the Play Games v2 sign-in screen is this one

![play_games_consent_screen](https://github.com/Abedalkareem/games_services/assets/113537099/a259ba18-c5fd-41df-8bc8-c94edc736860)

To the best of my knowledge, no other screen (that is, no consent screen) shows up after that. Which is great! Easy, safe, frictionless experience for the user.

- the Play Games v1 sign-in screen is the following one:

![playgamesv1signin](https://github.com/Abedalkareem/games_services/assets/113537099/bc286846-9c37-442b-bcc6-042036d5b141)

This is not as pretty as the Play Games v2 sign-in screen, but remains elegant.

So far so good.

However, there is a (scary) **consent screen** that follows after that, the same as when using `google_sign_in`:

![google_sign_in_games_consent_screen](https://github.com/Abedalkareem/games_services/assets/113537099/407d2d2a-b49f-4d80-90fb-73d11ef85b2c)

I am using `GamesServices.signIn(shouldEnableSavedGames = false)` (default option).

Do you confirm this consent screen (that does not show up with Play Games v2) is normal and expected? Remember, that's important because, as far as I know, the Play Games v2 SDK does not bother the user with this consent screen.

**Comment by andynewman10** [comment]
Ok, thanks for your investigations.

I said something wrong in the above thread, my mistake: the Play Games sign-in dialog is actually the same for both PGSv1 and PGSv2.

The sign-in dialog below seems to be displayed only when you have NEVER created a Play Games profile before:

![play_games_consent_screen](https://github.com/Abedalkareem/games_services/assets/113537099/a259ba18-c5fd-41df-8bc8-c94edc736860)

**Comment by andynewman10** [comment]
> Silly me should have pointed this out already, but even when using GoogleSignIn, it has to be used alongside Play Games Services. Therefore, even if you use GoogleSignIn, which yes, will always look different, you have to pass the credential on to the PGS SDK to activate Play Games features. Therefore, you still wouldn't get around integrating PGS, which as of v2 will automatically login without disabling the great mystery that is `PlayGamesInitProvider`.
> 
> So to answer question number 2, it will not give that seamless experience, and it will not eliminate the need for the Play Games SDK provided by games_services.

Understood, and I have verified it this morning too.

It is possible to view Google Account permissions using the following link: https://myaccount.google.com/connections?filters=3

When PGSv2 is used:

- signIn() is called
- the PGS sign-in dialog appears, you click Authorize.
- the following permissions are _automatically_ added to your account:
  - See, create, and delete its own configuration data in your Google Drive pentagon
  - Create, edit, and delete your Google Play Games activity

When PGSv1 is used:

- signIn() is called
- the PGS sign-in dialog appears, you click Authorize.
- the following permissions are _manually_ added, through a consent screen, to your account:
  - See, create, and delete its own configuration data in your Google Drive pentagon
  - Create, edit, and delete your Google Play Games activity
  - See your profile info <-- this seems to be the one which makes a difference

Conclusion: PGSv2 indeed has less friction.

**Comment by andynewman10** [comment]
I found how to bloody get rid of the Play Games bubble popup bogus behavior.

What we want is

- have the PG popup displayed when `signIn()` is called and only then
- not have it displayed when the app is brought back to the foreground

(We can live without signOut() - as long as the popup is not displayed when the player believes it is signed out from the game's perspective)

For this:

- the playgamesinitprovider needs to be disabled and the Auth.kt code needs to be changed as per the instructions I gave above, with a minor addition:

```
import android.app.Instrumentation

...

  fun signIn(activity: Activity?, result: MethodChannel.Result) {
    if (initialized == false) {
      Log.i("PlayService", "Initializing Play Games")
      PlayGamesSdk.initialize(activity!!) //  as Context
      initialized = true
      // Force the ActivityCallbacks to be called
      instrum.callActivityOnPause(activity!!)
      instrum.callActivityOnResume(activity!!)
    }
...
```
I checked, and this does the job. I don't expect `Instrumentation` to need some kind of specific permission, so we should be all set! I don't believe this requires to be added into the `games_services` code base, but at least this is a useful addition to know about.

**Comment by theLee3** [comment]
Awesome! I'm glad you got it sorted.

You should be good. There should be no needed permission.

**Comment by andynewman10** [comment]
I cannot thank you enough for your help @theLee3.

I am now closing this issue (I hope to complete the issue on issuetracker for further investigation on the Play Games' team side, I hope I'll find some more time for this).

**Comment by erperejildo** [comment]
> * signIn

I have a couple of question:

- Why, if I'm using `games_services: ^4.0.1` I see v1 login screen and no v2?
- If I'm not calling `signIn()`, why am I showing all the time the login screen on every build after sing in?

**Comment by andynewman10** [comment]
@erperejildo what I called the "Play Games v2" dialog is actually a dialog that is displayed when have never created a Play Games profile before. There is no v1/v2 difference (unless my memory is failing).
To reproduce, set up another phone, with a Google account where Play Games was never set up, do not install Play Games or anything on your phone and don't go in the Play Games settings (Android 12+), and try again. Because you have never created a profile before, you should see this dialog.

**Comment by erperejildo** [comment]
and why is it showing all the time? You mentioned here is it shown just once.

more info: https://github.com/Abedalkareem/games_services/issues/129#issuecomment-1999277650
